packages/component-library/components/Heading/Heading.elm

```diff
@@ -0,0 +1,270 @@
+module Heading.Heading exposing
+    ( Config
+    , DataAttribute
+    , TypeVariant(..)
+    , a
+    , addAttribute
+    , classNameAndIHaveSpokenToDST
+    , div
+    , h1
+    , h2
+    , h3
+    , h4
+    , h5
+    , h6
+    , id
+    , p
+    , pre
+    , span
+    , tag
+    , variant
+    , view
+    )
+
+import CssModules exposing (css)
+import Debug exposing (log)
+import Html exposing (text)
+import Html.Attributes
+import List
+
+
+type Config msg
+    = Config (ConfigValue msg)
+
+
+type alias Element msg =
+    List (Html.Attribute msg) -> List (Html.Html msg) -> Html.Html msg
+
+
+type alias DataAttribute =
+    { name : String
+    , value : String
+    }
+
+
+type alias ConfigValue msg =
+    { tag : Maybe (Element msg)
+    , variant : TypeVariant
+    , id : Maybe String
+    , data : List DataAttribute
+    , classNameAndIHaveSpokenToDST : Maybe String
+    }
+
+
+defaultConfig : ConfigValue msg
+defaultConfig =
+    { tag = Nothing
+    , variant = Display0
+    , id = Nothing
+    , data = []
+    , classNameAndIHaveSpokenToDST = Nothing
+    }
+
+
+toUntypedAttribute : DataAttribute -> Html.Attribute msg
+toUntypedAttribute typedAttribute =
+    case typedAttribute of
+        { name, value } ->
+            Html.Attributes.attribute ("data-" ++ name) value
+
+
+inferTagFromVariant : TypeVariant -> Element msg
+inferTagFromVariant typeVariant =
+    case typeVariant of
+        Display0 ->
+            Html.h1
+
+        Heading1 ->
+            Html.h1
+
+        Heading2 ->
+            Html.h2
+
+        Heading3 ->
+            Html.h3
+
+        Heading4 ->
+            Html.h4
+
+        Heading5 ->
+            Html.h5
+
+        Heading6 ->
+            Html.h6
+
+
+
+-- VIEW
+
+
+view : Config msg -> List (Html.Html msg) -> Html.Html msg
+view (Config config) children =
+    let
+        resolveId =
+            case config.id of
+                Just id_ ->
+                    [ Html.Attributes.id id_ ]
+
+                Nothing ->
+                    []
+
+        resolveAttributes =
+            List.map toUntypedAttribute config.data
+
+        resolveTag =
+            case config.tag of
+                Just tag_ ->
+                    tag_
+
+                Nothing ->
+                    inferTagFromVariant config.variant
+
+        resolveCustomClass =
+            case config.classNameAndIHaveSpokenToDST of
+                Just classNameAndIHaveSpokenToDST_ ->
+                    classNameAndIHaveSpokenToDST_
+
+                Nothing ->
+                    ""
```

Comment:
Surely there's an easier way to default a property that might not exist to an empty string.... right?

Reply:
I'm curious about this too – always do it the way you did here.

Reply:
You can use `Maybe.withDefault`:

It would look like this:
```
resolveCustomClass =
    config.classNameAndIHaveSpokenToDST |> Maybe.withDefault ""
```

or 

```
resolveCustomClass =
    Maybe.withDefault "" config.classNameAndIHaveSpokenToDST
```
(I like the pipeline version better even though it's a couple of extra chars)

Either of these two are certainly a lot shorter. But are they better? I dunno - I really don't mind explicit case matches on `Maybe`s (I think it reads a little better),

Reply:
TIL, thanks @mbylstra. I've updated to use `Maybe.withDefault` - much cleaner